---
title: "Analyze Collinsella"
author: "Yixuan Yang"
output:
  html_document:
    toc: true
    toc_float: true
    self_contained: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Set Up

```{r message=FALSE}
set.seed(2024)
library(tidyverse)
library(here)
library(rstatix)
library(phyloseq)
library(gt)
theme_set(theme_bw())

ps.rare <- readRDS(here('data','following_study','ps_rarefied.rds')) %>%
    filter_taxa(function(x) sum(x)>=50&sum(x>0)/length(x)>=0.1, prune = TRUE) # filter out taxa with prevalence lower than 10%
```

# Top Candidate

## Collinsella

We have found that `Collinsella` is the top candidate that influence dog's microbiome. Here we investigate its existence. In this section, we will mainly test the association between the existence of `Collinsella` and the epilepsy status of the dogs.

```{r}
ps.collinsella <- ps.rare %>%
    subset_taxa(Genus == 'Collinsella') 

# add genus level collinsella count to otu table
collinsella.df <- otu_table(ps.collinsella) %>% 
    cbind(Genus = rowSums(otu_table(ps.collinsella))) %>%
    cbind(data.frame(sample_data(ps.collinsella))) %>% 
    pivot_longer(c(starts_with('ASV'), 'Genus'), names_to = 'ASV', values_to = 'Abundance')


lfc <- collinsella.df %>%
    group_by(Household, ASV) %>%
    summarise(LFC = log2(Abundance[Epileptic.or.Control == "Epileptic"] + 0.5) -
                  log2(Abundance[Epileptic.or.Control == "Control"] + 0.5),
              .groups = 'drop')

ggplot(lfc) +
    geom_bar(aes(x = as.numeric(Household), y = LFC, fill = (LFC > 0)), stat = 'identity') +
    facet_wrap(~ASV, scales = 'free_y') +
    labs(x = 'Household', y = 'log 2 fold change', fill = 'LFC > 0')
```

From the previous plot, we can see that some of the Collinsella ASVs are more abundant in the epileptic group than the control group. We can test this hypothesis using the pairwise sign test.

```{r}
collinsella.df %>% 
    group_by(ASV) %>%
    arrange(Household) %>% # make sure data are paired
    pairwise_sign_test(Abundance ~ Epileptic.or.Control, ref.group = 'Epileptic', alternative = 'greater')
```

```{r}
ftable(xtabs(sign(Abundance) ~ Epileptic.or.Control + ASV, data = collinsella.df))
```

Previous studies shows that male dogs have a higher rate of having epilepsy. Here we investigate whether the existence of Collinsella is associated to sex.

```{r}
ftable(xtabs(sign(Abundance) ~ Sex  + Epileptic.or.Control + ASV, data = collinsella.df))
```

Chi-squared test between sex and health condition group for each ASV.

```{r}
collinsella.df %>% 
    split(.$ASV) %>% 
    lapply(function(df) summary(xtabs(sign(Abundance) ~ Sex  + Epileptic.or.Control, data = df)))
```

```{r}
ftable(xtabs(sign(Abundance) ~ Seizure.Freedom..Y.N.  + Pheno.Y.N + ASV, data = collinsella.df))
```

For epileptic dogs, does drug usage associate on sex to the existence frequency of Collinsella?

```{r warning=FALSE}
collinsella.df %>% 
    filter(Epileptic.or.Control == 'Epileptic') %>% 
    split(.$ASV) %>% 
    lapply(function(df) summary(xtabs(sign(Abundance) ~ Pheno.Y.N + Seizure.Freedom..Y.N., data = df)))
```

Does age associate to the abundance of Collinsella?

```{r}
collinsella.df %>%
    split(.$ASV) %>%
    lapply(function(df) MASS::glm.nb((Abundance + 1) ~ Age..months., data = df) %>% summary)
```

# Callback to Pilot Study

In the pilot study, we have found Lactobacillus was associated with epilepsy. Here we investigate them like we did above.

## Lactobacillus

```{r}
ps.lactobacillus <- ps.rare %>%
    subset_taxa(Genus == 'Lactobacillus') 

# add genus level lactobacillus count to otu table
lactobacillus.df <- otu_table(ps.lactobacillus) %>% 
    cbind(Genus = rowSums(otu_table(ps.lactobacillus))) %>%
    cbind(data.frame(sample_data(ps.lactobacillus))) %>% 
    pivot_longer(c(starts_with('ASV'), 'Genus'), names_to = 'ASV', values_to = 'Abundance')


lfc <- lactobacillus.df %>%
    group_by(Household, ASV) %>%
    summarise(LFC = log2(Abundance[Epileptic.or.Control == "Epileptic"] + 1) -
                  log2(Abundance[Epileptic.or.Control == "Control"] + 1),
              .groups = 'drop')

ggplot(lfc) +
    geom_bar(aes(x = as.numeric(Household), y = LFC, fill = (LFC > 0)), stat = 'identity') +
    facet_wrap(~ASV, scales = 'free_y') +
    labs(x = 'Household', y = 'log 2 fold change', fill = 'LFC > 0')
```

```{r}
lactobacillus.df %>% 
    group_by(ASV) %>%
    arrange(Household) %>% # make sure data are paired
    pairwise_sign_test(Abundance ~ Epileptic.or.Control, ref.group = 'Epileptic', alternative = 'greater')
# calculate prevalence
lactobacillus.df %>% 
    group_by(ASV) %>% 
    summarise(Prevalence = sum(Abundance > 0) / n())
```

```{r}
ftable(xtabs(sign(Abundance) ~ Epileptic.or.Control + ASV, data = lactobacillus.df))
```

Previous studies shows that male dogs have a higher rate of having epilepsy. Here we investigate whether the existence of lactobacillus is associated to sex.

```{r}
ftable(xtabs(sign(Abundance) ~ Sex + Epileptic.or.Control + ASV, data = lactobacillus.df))
```

Chi-squared test between sex and health condition group for each ASV.

```{r}
lactobacillus.df %>% 
    split(.$ASV) %>% 
    lapply(function(df) summary(xtabs(sign(Abundance) ~ Sex  + Epileptic.or.Control, data = df)))
```

```{r}
ftable(xtabs(sign(Abundance) ~ Sex  + Pheno.Y.N + ASV, data = lactobacillus.df))
```

For epileptic dogs, does drug usage associate on sex to the existence frequency of lactobacillus?

```{r}
lactobacillus.df %>% 
    filter(Epileptic.or.Control == 'Epileptic') %>% 
    split(.$ASV) %>% 
    lapply(function(df) summary(xtabs(sign(Abundance) ~ Pheno.Y.N, data = df)))
```

Does age associate to the abundance of lactobacillus?

```{r}
lactobacillus.df %>%
    split(.$ASV) %>%
    lapply(function(df) MASS::glm.nb((Abundance + 1) ~ Age..months., data = df) %>% summary)
```

# Taxonomy Information

```{r}
prev <- ps.rare %>% 
    psmelt() %>% 
    filter(Genus %in% c('Collinsella', 'Lactobacillus')) %>%
    group_by(OTU, Genus) %>%
    summarise(Frequency = sum(Abundance > 0),
              Prevalence = sum(Abundance > 0) / n()) %>% 
    arrange(Genus)
prev
tax_table(ps.rare)[prev$OTU, c('Genus','Species')]
```

Some of the species were not identified. We can use blast to search against the nt/nr database.

```{r}
# extract sequence of these ASVs
refseq(ps.rare)[prev$OTU] %>% dada2:::pfasta(ids = prev$OTU)
```

Here's the result:

| taxon | species | identity |
|:------|:--------|:---------|
| ASV2 | Collinsella stercoris | 100% |
| ASV44 | Collinsella aerofaciens | 100% |
| ASV70 | Collinsella tanakaei | 100% |
| ASV79 | Collinsella tanakaei | 98.81% |
| ASV83 | Collinsella phocaeensis | 100% |
| ASV105 | Lactobacillus gasseri | 100% |
| ASV55 | Lactobacillus crispatus | 100% |

```{r}
df <- data.frame(sample_data(ps.rare))
df.sub <- df %>% filter(Seizure.Control..Satisfactory.Unsatisfactory. == 'S')
xtabs(~ Pheno.Y.N+Seizure.Freedom..Y.N., data = df.sub) %>% summary()
```

