---
title: "Metadata Cleaning"
author: "Yixuan Yang"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    self_contained: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r message=FALSE}
rm(list = ls())
library(tidyverse)
library(readxl)
library(here)
library(phyloseq)
library(DECIPHER)
```

# Read in metadata

```{r}
file <- 'Epilepsy study mapping file with patient info updated 1.22.24 for microbiome analysis.xlsx'
meta_data <- read_excel(here('data', 'following_study', file), 
                        n_max = 118,
                        na = '.') %>%
    arrange(str_rank(Household, numeric = T))
# the sample id2 column was in float, we we change it to integer
str(meta_data)
meta_data$Household <- meta_data$Household %>% 
    as.character()
meta_data$`Age (months)` <- meta_data$`Age (months)` %>%
    str_remove('\\.0') %>% 
    as.numeric()
meta_data$`Zip Code` <- meta_data$`Zip Code` %>%
    str_remove('\\.0') %>%
    str_replace('O','0')
# remove comments row in excel
meta_data <- meta_data %>% dplyr::select(-COMMENTS) %>% data.frame()
# match sample name same as that in the asv table
rownames(meta_data) <- 'Netti' %>%
    str_c(sprintf("%03d", as.integer(meta_data$SampleID2)))
meta_data <- meta_data %>% add_column(study = 'following')
meta_data <- meta_data %>% 
    arrange(str_rank(Household, numeric = T), Epileptic.or.Control)
```

Sex column contains information about reproductive status, here we separate them into two columns

```{r}
meta_data <- meta_data %>%
    separate(Sex, into = c("Sex", "Reproductive.Status"), sep = 1) %>% 
    mutate(Spayed.or.Neutered = (Reproductive.Status != 'I'))
```

Pheno.Y.N. indicates whether the dogs are exposed to phenobarbital or not. Here assign NA to No.

```{r}
meta_data$Pheno.Y.N[is.na(meta_data$Pheno.Y.N)] <- 'No'
```


# Create phyloseq object

Store the metadata, sequence table, and taxonomy table into a phyloseq object

```{r}
st <- readRDS(here('data','following_study','st.rds')) # sequence table
taxa <- readRDS(here('data','following_study','taxsp.rds')) # taxonomy table

# match st order with metadata (since metadata is sorted by household)
st <- st[meta_data$SampleID2,]
if(!identical(nrow(meta_data), nrow(st))) stop('meta data does not match with sequence table')
if(!identical(rownames(meta_data), rownames(st))) stop('rownames in meta data does not match with sequence table')
# create a phyloseq object
ps <- phyloseq(sample_data(meta_data),
               otu_table(st, taxa_are_rows = FALSE),
               tax_table(taxa))

# Add ASV sequences
seqs <- Biostrings::DNAStringSet(taxa_names(ps))
names(seqs) <- taxa_names(ps)
ps <- merge_phyloseq(ps, seqs)
taxa_names(ps) <- paste0("ASV", seq(ntaxa(ps))) 
```

# Quality Assurance

Do some QA on the sequencing libraries:

```{r}
plot(rowSums(st), log="y")
which(rowSums(st) < 2e+3)
which(rowSums(st) < 100)
```

All of the data has more than 100 reads, but sample 40,80 and 108 have relative low sequencing depth.

# Phylogenetic Tree construction

Here, we use DECIPHER package to align sequence and then construct phylogenetic tree using maximum likelihood method with `GTR+G4` model.

```{r, eval = FALSE}
set.seed(2024)
alignment <- AlignSeqs(refseq(ps), processors=NULL) 
tree <- Treeline(alignment, processors=NULL)
WriteDendrogram(tree, file=here('data','following_study','tree.nwk'))
```

add phylogenetic information to pyholseq objects

```{r}
tree <- read_tree(here('data','following_study','tree.nwk'))
ps <- merge_phyloseq(ps, phy_tree(tree))
plot_tree(ps, method = 'treeonly')
```

# Data Cleaning

Some dogs have duplicated samples. Here, we keep the one with the most reads.

In household 52, sample 106,107,108 are duplicates of sample 103.

```{r}
sort(rowSums(st)[str_c('Netti',c('103','106','107','108'))])
```

sample 105 is the duplicate of sample 104.

```{r}
sort(rowSums(st)[str_c('Netti',c('104','105'))])
```

So, for household 52, sample 103,104 are kept.

In household 54, sample 63,113,115 are duplicates of sample 114. 

```{r}
rowSums(st)[str_c('Netti',c('063','113','114','115'))] %>% sort()
```

sample 64,117,118 are duplicates of sample 116.

```{r}
rowSums(st)[str_c('Netti',c('064','116','117','118'))] %>% sort()
```

So, for household 52, sample 64,114 are kept.

now we can remove sample 106,107,108,113,115,63,116,117,118 since they are duplicates with lower sequencing depth.

```{r}
duplicated.samples <- c(104,106,107,108,113,115,63,116,117,118)
ps <- subset_samples(ps, !(SampleID2 %in% duplicated.samples))
```

Some of the households have more than 2 samples, we force to keep two dogs to maintain the structure of the study design.

```{r}
n.samples <- table(sample_data(ps)$Household)
n.samples[n.samples != 2]
```

There are 4 dogs in household 17, 18, and 26.

For household 17, we keep Netti 109 and 112. 

```{r}
sample_data(ps) %>%
    data.frame() %>% 
    mutate(read.depth = rowSums(otu_table(ps))) %>% 
    filter(Household == 17) %>%
    dplyr::select(SampleID2, Household, Epileptic.or.Control, read.depth)
```

For household 18, we keep Netti 049 and 050.

```{r}
sample_data(ps) %>%
    data.frame() %>% 
    mutate(read.depth = rowSums(otu_table(ps))) %>% 
    filter(Household == 18) %>%
    dplyr::select(SampleID2, Household, Epileptic.or.Control, read.depth)
```

For household 26, we keep Netti 025 and 026

```{r}
sample_data(ps) %>%
    data.frame() %>% 
    mutate(read.depth = rowSums(otu_table(ps))) %>% 
    filter(Household == 26) %>%
    dplyr::select(SampleID2, Household, Epileptic.or.Control, read.depth)
```


Remove exceeded samples to make a paired data.

```{r}
remove.list <- c(110,111,9,10,37,38)
ps <- subset_samples(ps, !(SampleID2 %in% remove.list))
ps
```

Now we have 102 samples (data are paired within 51 households)

# Rarefied Data

The sequencing depth for each sample is different, we need to check and bring them to the same level.

```{r}
# check distribution of sequencing depth
sample_sums(ps) %>% hist()
sample_sums(ps) %>% sort() %>% head()
```

Here, we see most of sample has sequencing depth greater than 40000, we can bring all the samples to this sequencing depth level, and drop the two samples (Netti040 and Netti080) that have low sequencing depth.


```{r message=FALSE}
# make read depth even
ps.rarefied <- rarefy_even_depth(ps, sample.size = 44590)
ps.rarefied
```

Two samples are lost after rarefying. To keep the consistency between the rarefied and non-rarefied data and the structure of the pairwise study design, we remove pairs from household 37 and 46.

```{r message=FALSE}
house.rm <- meta_data[c('Netti040','Netti080'),'Household']
house.rm
```

```{r}
ps <- subset_samples(ps, !(Household %in% house.rm))
# since some households removed, remove taxa that are not present in the remained sample
ps <- filter_taxa(ps, function (x) sum(x) > 0, prune = TRUE)
ps
```

```{r}
ps.rarefied <- subset_samples(ps.rarefied, !(Household %in% house.rm))
ps.rarefied
```

check if data are paired by household

```{r}
if (!identical(sample_data(ps)$Household[seq(1,98,2)], sample_data(ps)$Household[seq(1,98,2)+1])) stop('data is not paired')
if (!identical(sample_data(ps.rarefied)$Household[seq(1,98,2)], sample_data(ps.rarefied)$Household[seq(1,98,2)+1])) stop('data is not paired')
```



# Save Data

check names are matched for both datasets

```{r}
if (!identical(rownames(otu_table(ps)), rownames(sample_data(ps)))) stop('sample names are not matched!')
if (!identical(rownames(otu_table(ps.rarefied)), rownames(sample_data(ps.rarefied)))) stop('sample names are not matched!')
```


```{r}
saveRDS(ps, file = here('data','following_study','ps.rds'))
saveRDS(ps.rarefied, file = here('data','following_study','ps_rarefied.rds'))
```

```{r}
sessioninfo::session_info()
```


