---
title: "analysis"
author: "Yixuan Yang"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
rm(list = ls())
library(tidyverse)
library(dada2)
theme_set(theme_bw())
```


```{r}
# load preprocessed data
st <- readRDS('../Rdata/st_new.rds')
tax <- readRDS('../Rdata/tax.rds')
taxsp <- readRDS('../Rdata/taxsp.rds')
meta_data <- readRDS('../Rdata/metadata.rds')
ft <- sweep(st, 1, rowSums(st), "/")

meta_data$Group <- as.factor(meta_data$Group)
```

Here we can first look at helicobactor.

```{r}
helico.sq <- getSequences(tax)[which(tax[,6] %in% 'Helicobacter')]
taxsp[helico.sq,7] %>% unname
```

Not all sequences are assigned to a species. Here we need to use web BLAST

```{r}
dada2:::pfasta(helico.sq)
```
```{r}
helico.species <- c('canis',
                    'canis',
                    'canis',
                    'canicola/cinaedi/bills',
                    'canicola/cinaedi/bills',
                    'canis',
                    'canis', # 99.6
                    'bills',
                    'canicola/cinaedi/bills',
                    'canis', # 99.6
                    'canicola/cinaedi',
                    'winghamensis',
                    'canis') # 95.56

canis.sq <- helico.sq[helico.species =='canis']
```

```{r}
canis.prop <- ft[,canis.sq]
colnames(canis.prop) <- str_c('helico.canis',1:7)
cor(canis.prop)
outer(canis.sq, canis.sq, nwhamming, vec = T)
```

from the correlation matrix, we can see a high correlation between helico.canis 4 and 5. It could be a intragenome variant. The correlation between helico.canis 3 and 7 is also relatively high.

from the difference matrix, we see the first 6 species are similar, but species # 7 has a large difference than the previous 6.


```{r}
meta_data.canis <- meta_data %>% cbind(canis.prop = rowSums(canis.prop))
```

```{r}
meta_data.canis <- meta_data.canis %>% arrange(str_rank(Household, numeric = TRUE))
meta_data.canis$Household %>% table()
```


```{r}
ggplot(meta_data.canis) +
  geom_boxplot(aes(x=Group, y=canis.prop))
ggplot(meta_data.canis) +
  geom_boxplot(aes(x=`Pheno or drug naive`, y=canis.prop))
```

```{r}
groupA <- meta_data.canis %>% filter(Group == 'A')
groupB <- meta_data.canis %>% filter(Group == 'B')
if(!identical(groupA$Household, groupB$Household)) stop ('Household mismatch')
t.test(groupA$canis.prop, groupB$canis.prop, paired = T, alternative = 'less')
t.test(data = meta_data, canis.prop~`Pheno or drug naive`, paired = F, alternative = 'greater')
```

```{r}
taxdf <- data.frame(tax)
freq.table <- (table(taxdf$Genus) %>% sort(decreasing = T))[1:100] %>%
    data.frame() %>%
    setNames(c('Genus', 'number of ASV'))

pv <- data.frame(p.value=c(), meanDiff=c())

for (genus in freq.table$Genus) {
  seq <- getSequences(tax)[which(tax[,6] %in% genus)]
  prop <- ft[,seq] %>% rowSums()
  df <- cbind(meta_data, prop)
  test.result <- t.test(prop~Group, data = df, paired = TRUE)
  pv <- pv %>% bind_rows(test.result[c('p.value', 'estimate')] %>% unlist())
}
pvalues <- freq.table %>% cbind(pv)

pvalues %>% filter(p.value <= 0.2) %>% arrange(p.value)
```







