---
title: "Pilot Study"
author: "Yixuan Yang"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Set Up

```{r message=FALSE}
library(tidyverse)
library(here)
library(phyloseq)
library(dada2)
library(DECIPHER)
```

# Read in metadata

```{r}
metadata <- read.csv(here('data','pilot_study','epilepsy_metadata.tsv'), sep = '\t') %>% 
    add_column(study = 'pilot')
rownames(metadata) <- metadata$SampleID
head(metadata)
```

# Read in sequence table

```{r}
st <- readRDS(here('data','pilot_study','st.rds'))
rownames(st) <- rownames(st) %>% str_remove('_F.fastq.gz')
if (!identical(rownames(metadata), rownames(st))) {stop('Sample names do not match')}
```

# Assign taxonomy using Silva 138.1

```{r eval=FALSE}
tax <- assignTaxonomy(st, "~/tax/silva_nr99_v138.1_train_set.fa.gz", multithread=TRUE)
taxsp <- addSpecies(tax, "~/tax/silva_species_assignment_v138.1.fa.gz")
saveRDS(tax, here("data", "pilot_study", "tax.rds"))
saveRDS(taxsp, here("data", "pilot_study", "taxsp.rds"))
```

# Save everything to phyloseq object

```{r}
taxsp <- readRDS(here("data", "pilot_study", "taxsp.rds"))
ps <- phyloseq(sample_data(metadata),
               otu_table(st, taxa_are_rows = FALSE), 
               tax_table(taxsp))
# Add ASV sequences
seqs <- Biostrings::DNAStringSet(taxa_names(ps))
names(seqs) <- taxa_names(ps)
ps <- merge_phyloseq(ps, seqs)
taxa_names(ps) <- paste0("ASV", seq(ntaxa(ps))) 
```

# Construct the phylogenetic tree

```{r eval=FALSE}
set.seed(2024)
alignment <- AlignSeqs(refseq(ps), processors=NULL) 
tree <- Treeline(alignment, method = "ME", processors=NULL)
WriteDendrogram(tree, file=here('data','pilot_study','tree.nwk'))
```

add phylogenetic information to pyholseq objects

```{r}
# tree <- read_tree(here('data','pilot_study','tree.nwk'))
# ps <- merge_phyloseq(ps, phy_tree(tree))
# plot_tree(ps, method = 'treeonly')
```

# Rarefied Data

The sequencing depth for each sample is different, we need to check and bring them to the same level.

```{r}
# check distribution of sequencing depth
sample_sums(ps) %>% sort() %>% head()
```

Here we see `020Petey` has very low sequencing depth, we will remove the pair and rarefy the data to 16683.

```{r}
sample_data(ps)['020Petey', 'Household']
ps <- subset_samples(ps, Household != '10')
```



```{r message=FALSE}
# make read depth even
ps.rarefied <- rarefy_even_depth(ps, sample.size = 16683)
```

# Save the phyloseq object

non-rarefied data

```{r}
ps
```

rarefied data

```{r}
ps.rarefied
```


```{r}
saveRDS(ps, file = here('data','pilot_study','ps.rds'))
saveRDS(ps.rarefied, file = here('data','pilot_study','ps_rarefied.rds'))
```


